psi4/driver/driver.py

```diff
@@ -873,6 +873,158 @@ def properties(*args, **kwargs):
         return core.variable('CURRENT ENERGY')
 
 
+
+def optimize_geometric(name, **kwargs):
+
+    import geometric
+    import qcelemental as qcel
+
+    class Psi4NativeEngine(geometric.engine.Engine):
+        """
+        Run a Psi4 energy and gradient calculation for geometric without any I/O
+        """
+        def __init__(self, p4_name, p4_mol, **p4_kwargs):
+    
+            self.p4_name = p4_name
+            self.p4_mol = p4_mol
+            self.p4_kwargs = p4_kwargs
+    
+            molecule = geometric.molecule.Molecule()
+            molecule.elem = [p4_mol.symbol(i) for i in range(p4_mol.natom())]
+            molecule.xyzs = [p4_mol.geometry().np * qcel.constants.bohr2angstroms] # reshape?
+            molecule.build_bonds()
+                                 
+            super(Psi4NativeEngine, self).__init__(molecule)
+    
+        def calc(self, coords, dirname):
+            self.p4_mol.set_geometry(core.Matrix.from_array(coords.reshape(-1,3)))
+            self.p4_mol.update_geometry()
+            g, wfn = gradient(self.p4_name, return_wfn=True, molecule=self.p4_mol, **self.p4_kwargs)
+            e = core.variable('CURRENT ENERGY')
+            return {'energy': e, 'gradient': (g.np).reshape(-1)}
+
+    if '/' in name:
+        raise ValidationError(f'Composite methods like {name} are not yet supported with geomeTRIC')
+                                 
+    return_wfn = kwargs.pop('return_wfn', False)
+
+    # Make sure the molecule the user provided is the active one
+    molecule = kwargs.get('molecule', core.get_active_molecule())
+
+    # Do not fix orientation or COM
+    molecule.fix_orientation(True)
+    molecule.fix_com(True)
+    molecule.update_geometry()
+
+    # Get geometric-specific options
+    geometric_opts = kwargs.get('geometric_opts', {})
+
+    # Default to Psi4 maxiter unless overridden
+    if 'maxiter' not in geometric_opts:
+        geometric_opts['maxiter'] = core.get_global_option('GEOM_MAXITER')
+
+    # Default to Psi4 geometry convergence criteria unless overridden 
+    if 'convergence_set' not in geometric_opts:
+        geometric_opts['convergence_set'] = core.get_global_option('G_CONVERGENCE')
+
+        # GeomeTRIC doesn't know these convergence criterion
+        if geometric_opts['convergence_set'] in ['CFOUR', 'QCHEM', 'MOLPRO']:
+            geometric_opts['convergence_set'] = 'GAU_TIGHT'
+
+    engine = Psi4NativeEngine(name, molecule, **kwargs)
+    M = engine.M
+    
+    # Handle constraints
+    constraints_dict = geometric_opts.get('constraints', {})
+    constraints_string = geometric.run_json.make_constraints_string(constraints_dict)
+    Cons, CVals = None, None
+    if constraints_string:
+        if 'scan' in constraints_dict:
+            raise ValueError("No scan!")
```

Comment:
A bit more verbose, like `"scan not yet implemented"` would be better. Why doesn't it work? Not implemented in GeomeTRIC, or in the Psi4-GeomeTRIC interface?

Reply:
As far as I know the scan feature works in GeomeTRIC, but I think that hooking it up in Psi4 would require additional accommodations to the geometry optimization loop and print out (since you're iterating over additional variables). I changed the print out to be a little more clear.